File: results/2026-05-06_173435/tasks/12-pr-label-assigner/default-opus47-1m-xhigh/generated-code/pr_label_assigner.py
#!/usr/bin/env python3
"""PR label assigner.

Given a list of changed file paths and a set of pattern->label rules, decide
which labels apply to the PR. Supports gitignore-style globs (* / ** / ?),
multiple labels per file, and group-scoped priority for conflict resolution.

Approach
--------
- Each rule has a `pattern`, a `label`, an optional `priority` (int, default 0),
  and an optional `group` (string).
- For every file we evaluate every rule. A rule with no `/` in its pattern is
  treated as a basename-anywhere match (we prepend `**/`); otherwise the
  pattern must match the full path.
- All matching rules contribute their (label, priority) pairs.
- Conflict resolution: within a single `group`, only the highest-priority
  matching label survives. Ungrouped rules never conflict with each other.
- Output is sorted by priority descending, then alphabetically — deterministic
  so test assertions can compare exact strings.
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from dataclasses import dataclass
from pathlib import Path


class LabelAssignerError(Exception):
    """Raised for any user-facing config or input error."""


@dataclass(frozen=True)
class _CompiledRule:
    pattern: str
    label: str
    priority: int
    group: str | None
    regex: re.Pattern[str]


def _glob_to_regex(pattern: str) -> re.Pattern[str]:
    """Convert a glob pattern to an anchored regex.

    Semantics:
    - `**` matches any sequence of characters including `/`.
    - `**/` matches zero or more directory components.
    - `*` matches anything but `/`.
    - `?` matches a single non-`/` character.
    - All other regex metacharacters are escaped literally.
    - A pattern with no slash is treated as `**/<pattern>` (basename-anywhere).
    """
    if "/" not in pattern:
        pattern = "**/" + pattern

    out: list[str] = []
    i = 0
    n = len(pattern)
    while i < n:
        c = pattern[i]
        if c == "*":
            if i + 1 < n and pattern[i + 1] == "*":
                # `**` — handle the common `**/` form specially so it matches
                # zero directories (e.g. `**/x.py` should match `x.py`).
                if i + 2 < n and pattern[i + 2] == "/":
                    out.append("(?:.*/)?")
                    i += 3
                else:
                    out.append(".*")
                    i += 2
            else:
                out.append("[^/]*")
                i += 1
        elif c == "?":
            out.append("[^/]")
            i += 1
        elif c in r".+()|^$\{}[]":
            out.append("\\" + c)
            i += 1
        else:
            out.append(c)
            i += 1

    return re.compile("^" + "".join(out) + "$")


class LabelAssigner:
    """Apply a set of pattern->label rules to a list of changed files."""

    def __init__(self, rules: list[dict]) -> None:
        self._rules: list[_CompiledRule] = []
        for idx, raw in enumerate(rules):
            if "pattern" not in raw:
                raise LabelAssignerError(f"rule {idx}: missing required field 'pattern'")
            if "label" not in raw:
                raise LabelAssignerError(f"rule {idx}: missing required field 'label'")
            try:
                regex = _glob_to_regex(raw["pattern"])
            except re.error as e:
                raise LabelAssignerError(
                    f"rule {idx}: invalid pattern {raw['pattern']!r}: {e}"
                ) from e
            self._rules.append(
                _CompiledRule(
                    pattern=raw["pattern"],
                    label=raw["label"],
                    priority=int(raw.get("priority", 0)),
                    group=raw.get("group"),
                    regex=regex,
                )
            )

    def assign_labels(self, files: list[str]) -> list[str]:
        """Return the deterministic, deduplicated list of labels for these files."""
        # Collect every (label, priority, group) match across all (file, rule) pairs.
        matches: list[tuple[str, int, str | None]] = []
        for path in files:
            for rule in self._rules:
                if rule.regex.match(path):
                    matches.append((rule.label, rule.priority, rule.group))

        # Group conflict resolution: within a group, keep only the label whose
        # rule had the highest priority. Ties broken alphabetically by label
        # so output stays deterministic.
        best_per_group: dict[str, tuple[int, str]] = {}
        ungrouped: dict[str, int] = {}
        for label, priority, group in matches:
            if group is None:
                # Ungrouped: keep the best priority observed for the same label.
                if label not in ungrouped or priority > ungrouped[label]:
                    ungrouped[label] = priority
            else:
                cur = best_per_group.get(group)
                if cur is None or priority > cur[0] or (priority == cur[0] and label < cur[1]):
                    best_per_group[group] = (priority, label)

        final: dict[str, int] = dict(ungrouped)
        for priority, label in best_per_group.values():
            # If the same label arrives via both grouped and ungrouped rules,
            # keep the higher priority.
            if label not in final or priority > final[label]:
                final[label] = priority

        # Sort: priority desc, then label asc.
        return [lbl for lbl, _ in sorted(final.items(), key=lambda kv: (-kv[1], kv[0]))]


def load_rules(path: Path | str) -> list[dict]:
    """Load and validate rules from a JSON config file.

    Expected shape: ``{"rules": [{"pattern": ..., "label": ..., ...}, ...]}``.
    """
    p = Path(path)
    if not p.exists():
        raise LabelAssignerError(f"rules file not found: {p}")
    try:
        data = json.loads(p.read_text())
    except json.JSONDecodeError as e:
        raise LabelAssignerError(f"invalid JSON in {p}: {e}") from e

    if not isinstance(data, dict) or "rules" not in data:
        raise LabelAssignerError(
            f"{p}: top-level object must contain a 'rules' array"
        )
    rules = data["rules"]
    if not isinstance(rules, list):
        raise LabelAssignerError(f"{p}: 'rules' must be an array")

    for idx, r in enumerate(rules):
        if not isinstance(r, dict):
            raise LabelAssignerError(f"{p}: rule {idx} must be an object")
        if "pattern" not in r:
            raise LabelAssignerError(f"{p}: rule {idx} missing 'pattern'")
        if "label" not in r:
            raise LabelAssignerError(f"{p}: rule {idx} missing 'label'")
    return rules


def load_files(path: Path | str) -> list[str]:
    """Load a list of changed file paths from JSON.

    Accepts either ``{"files": [...]}`` or a bare JSON array.
    """
    p = Path(path)
    if not p.exists():
        raise LabelAssignerError(f"files file not found: {p}")
    try:
        data = json.loads(p.read_text())
    except json.JSONDecodeError as e:
        raise LabelAssignerError(f"invalid JSON in {p}: {e}") from e
    if isinstance(data, dict) and "files" in data:
        files = data["files"]
    elif isinstance(data, list):
        files = data
    else:
        raise LabelAssignerError(
            f"{p}: expected either an array or an object with a 'files' array"
        )
    if not all(isinstance(f, str) for f in files):
        raise LabelAssignerError(f"{p}: every entry in 'files' must be a string")
    return files


def main(argv: list[str] | None = None) -> int:
    """CLI entry point. Returns process exit code."""
    parser = argparse.ArgumentParser(
        description="Assign PR labels based on changed file paths.",
    )
    parser.add_argument(
        "--rules", required=True,
        help="Path to a JSON file with the rule set.",
    )
    parser.add_argument(
        "--files", required=True,
        help="Path to a JSON file listing the PR's changed files.",
    )
    parser.add_argument(
        "--format", choices=("lines", "json"), default="lines",
        help="Output format: 'lines' (one label per line) or 'json'.",
    )
    args = parser.parse_args(argv)

    try:
        rules = load_rules(args.rules)
        files = load_files(args.files)
        labels = LabelAssigner(rules).assign_labels(files)
    except LabelAssignerError as e:
        print(f"error: {e}", file=sys.stderr)
        return 2

    if args.format == "json":
        print(json.dumps({"labels": labels}))
    else:
        for lbl in labels:
            print(lbl)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: results/2026-05-06_173435/tasks/12-pr-label-assigner/typescript-bun-opus47-1m-xhigh/generated-code/pr-label-assigner.ts
// PR Label Assigner — applies labels to a list of changed file paths
// according to a configurable set of glob-pattern -> label rules.
//
// Public surface:
//   - Config / Rule / ExclusiveGroup types
//   - assignLabels(config, files): the pure logic
//   - main()  : CLI entry that reads --config and --files, prints JSON
//
// Design notes:
//   - Uses Bun.Glob for pattern matching against path strings.
//   - Each rule has one label and one or more glob patterns.
//   - A file may match many rules (multiple labels per file).
//   - Optional `exclusiveGroups` resolve conflicts between mutually
//     exclusive labels by keeping the highest-priority winner only.
//   - Output is the deduplicated label set, sorted by priority desc
//     (and label name asc as a tiebreaker) for stable, comparable output.

export interface Rule {
  label: string;
  patterns: string[];
  priority?: number;
}

export interface ExclusiveGroup {
  labels: string[];
}

export interface Config {
  rules: Rule[];
  exclusiveGroups?: ExclusiveGroup[];
}

/**
 * Pure function: given a config and a list of changed file paths,
 * return the final, sorted set of labels.
 *
 * Steps:
 *   1. For every rule, check whether any pattern matches any file.
 *      If so, the rule contributes its label.
 *   2. Resolve `exclusiveGroups`: within each group, only the
 *      highest-priority matched label survives (alphabetical tiebreak).
 *   3. Sort the surviving label set by priority desc, then label asc.
 */
export function assignLabels(config: Config, files: string[]): string[] {
  // Index rule priorities by label so we can sort & resolve later.
  const priorityOf = new Map<string, number>();
  for (const rule of config.rules) {
    priorityOf.set(rule.label, rule.priority ?? 0);
  }

  const matched = new Set<string>();
  for (const rule of config.rules) {
    const globs = rule.patterns.map((p) => new Bun.Glob(p));
    if (files.some((f) => globs.some((g) => g.match(f)))) {
      matched.add(rule.label);
    }
  }

  // Resolve exclusive groups by keeping only the highest-priority winner.
  for (const group of config.exclusiveGroups ?? []) {
    const present = group.labels.filter((l) => matched.has(l));
    if (present.length <= 1) continue;
    const winner = present.sort(comparator(priorityOf))[0]!;
    for (const l of present) {
      if (l !== winner) matched.delete(l);
    }
  }

  return [...matched].sort(comparator(priorityOf));
}

// Sort: priority desc, then label name asc (stable, deterministic order).
function comparator(priorityOf: Map<string, number>) {
  return (a: string, b: string): number => {
    const dp = (priorityOf.get(b) ?? 0) - (priorityOf.get(a) ?? 0);
    if (dp !== 0) return dp;
    return a < b ? -1 : a > b ? 1 : 0;
  };
}

// --- CLI ---------------------------------------------------------------

interface CliArgs {
  configPath: string;
  filesPath: string;
}

function parseArgs(argv: string[]): CliArgs {
  let configPath = "";
  let filesPath = "";
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--config") configPath = argv[++i] ?? "";
    else if (a === "--files") filesPath = argv[++i] ?? "";
    else throw new Error(`Unknown argument: ${a}`);
  }
  if (!configPath) throw new Error("Missing required --config <path>");
  if (!filesPath) throw new Error("Missing required --files <path>");
  return { configPath, filesPath };
}

async function loadConfig(path: string): Promise<Config> {
  const file = Bun.file(path);
  if (!(await file.exists())) {
    throw new Error(`config file not found: ${path}`);
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(await file.text());
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new Error(`invalid JSON in config ${path}: ${msg}`);
  }
  validateConfig(parsed);
  return parsed;
}

function validateConfig(c: unknown): asserts c is Config {
  if (!c || typeof c !== "object") {
    throw new Error("config must be an object");
  }
  const obj = c as Record<string, unknown>;
  if (!Array.isArray(obj.rules)) {
    throw new Error("config.rules must be an array");
  }
  for (const [i, r] of obj.rules.entries()) {
    if (!r || typeof r !== "object") {
      throw new Error(`config.rules[${i}] must be an object`);
    }
    const rule = r as Record<string, unknown>;
    if (typeof rule.label !== "string" || rule.label.length === 0) {
      throw new Error(`config.rules[${i}].label must be a non-empty string`);
    }
    if (
      !Array.isArray(rule.patterns) ||
      !rule.patterns.every((p) => typeof p === "string")
    ) {
      throw new Error(`config.rules[${i}].patterns must be a string[]`);
    }
    if (rule.priority !== undefined && typeof rule.priority !== "number") {
      throw new Error(`config.rules[${i}].priority must be a number`);
    }
  }
  if (obj.exclusiveGroups !== undefined && !Array.isArray(obj.exclusiveGroups)) {
    throw new Error("config.exclusiveGroups must be an array");
  }
}

async function loadFiles(path: string): Promise<string[]> {
  const file = Bun.file(path);
  if (!(await file.exists())) {
    throw new Error(`files list not found: ${path}`);
  }
  return (await file.text())
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter((l) => l.length > 0);
}

async function main(argv: string[]): Promise<number> {
  try {
    const args = parseArgs(argv);
    const [config, files] = await Promise.all([
      loadConfig(args.configPath),
      loadFiles(args.filesPath),
    ]);
    const labels = assignLabels(config, files);
    process.stdout.write(JSON.stringify({ labels }) + "\n");
    return 0;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    process.stderr.write(`error: ${msg}\n`);
    return 1;
  }
}

// Run as CLI when invoked as the main module.
if (import.meta.main) {
  process.exit(await main(process.argv.slice(2)));
}
